hack/ci/setup-kind-cluster.sh

```diff
@@ -0,0 +1,232 @@
+
+# Create kind cluster
+TEST_NAME="Create kind cluster"
+
+echodate "Preloading the kindest/node image"
+docker load --input /kindest.tar
+
+echodate "Creating the kind cluster"
+export KUBECONFIG=~/.kube/config
+
+beforeKindCreate=$(nowms)
+
+cat << EOF > kind-config.yaml
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+name: "${KIND_CLUSTER_NAME}"
+networking:
+  apiServerAddress: "0.0.0.0"
+  disableDefaultCNI: true # disable kindnet
+kubeadmConfigPatches:
+- |
+  kind: ClusterConfiguration
+  apiServer:
+    extraArgs:
+      "kubernetes-service-node-port": "31443"
+    certSANs:
+      - localhost
+      - 127.0.0.1
+      - kubernetes
+      - kubernetes.default.svc
+      - kubernetes.default.svc.cluster.local
+      - 0.0.0.0
+      - ${NODE_IP}
+      - ${KIND_CLUSTER_NAME}
+nodes:
+  - role: control-plane
+EOF
+
+if [ -n "${DOCKER_REGISTRY_MIRROR_ADDR:-}" ]; then
+  mirrorHost="$(echo "$DOCKER_REGISTRY_MIRROR_ADDR" | sed 's#http://##' | sed 's#/+$##g')"
+
+  # make the registry mirror available as a socket,
+  # so we can mount it into the kind cluster
+  mkdir -p /mirror
+  socat UNIX-LISTEN:/mirror/mirror.sock,fork,reuseaddr,unlink-early,mode=777 TCP4:$mirrorHost &
+
+  function end_socat_process {
+    echodate "Killing socat docker registry mirror processes..."
+    pkill -e socat
+  }
+  appendTrap end_socat_process EXIT
+
+  cat << EOF | tee -a /etc/environment
+    # mount the socket
+    extraMounts:
+    - hostPath: /mirror
+      containerPath: /mirror
+containerdConfigPatches:
+  # point to the soon-to-start local socat process
+  - |-
+    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
+    endpoint = ["http://127.0.0.1:5001"]
+EOF
+
+  kind create cluster --config kind-config.yaml
+  pushElapsed kind_cluster_create_duration_milliseconds $beforeKindCreate
+
+  # unwrap the socket inside the kind cluster and make it available on a TCP port,
+  # because containerd/Docker doesn't support sockets for mirrors.
+  docker exec $KIND_CLUSTER_NAME-control-plane bash -c 'socat TCP4-LISTEN:5001,fork,reuseaddr UNIX:/mirror/mirror.sock &'
+else
+  kind create cluster --config kind-config.yaml
+fi
+
+echodate "Kind cluster $KIND_CLUSTER_NAME is up and running."
+
+if [ ! -f cni-plugin-deployed ]; then
+  echodate "Installing CNI plugin."
+  (
+    # Install CNI plugins since they are not installed by default in KIND. Also, kube-flannel doesn't install
+    # CNI plugins unlike other plugins so we have to do it manually.
+    setup_cni_in_kind=$(cat hack/ci/setup-cni-in-kind.sh)
+    docker exec $KIND_CLUSTER_NAME-control-plane bash -c "$setup_cni_in_kind &"
+  )
+
+  kubectl create -f https://raw.githubusercontent.com/flannel-io/flannel/v0.18.0/Documentation/kube-flannel.yml
+  touch cni-plugin-deployed
+fi
+
+if [ -z "${DISABLE_CLUSTER_EXPOSER:-}" ]; then
+  # Annotate kube-apiserver service so that the cluster exposer can expose it
+  kubectl annotate svc kubernetes -n default nodeport-proxy.k8s.io/expose=true
```

Comment:
If we still use the clusterexposer and a NodePort service, why do we need to figure out the node IP where the Prow pod is running on?

Reply:
We need that node IP and the associated node port to find the external address for the kube-apiserver.